### ISSUE TITLE
[API Proposal] Add JIT Metrics APIs

### ISSUE THREAD

**Issue by josalem** [dotnet/runtime#54444]
## Background and Motivation

Similar to GC metrics, a user may want to quickly inspect information about the JIT while their application is running. This includes information like the number of methods that have been JIT'd, the number of bytes of code generated by the JIT, and time spent compiling code in the JIT. We already produce `EventCounter`s for `il-bytes-jitted` and `methods-jitted-count` (see: https://github.com/dotnet/runtime/blob/8761cb9cf2201054409dd8e0bc2e6bfd85ae22bd/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs#L86-L87). The APIs queried for those counters are currently internal (see: https://github.com/dotnet/runtime/blob/8761cb9cf2201054409dd8e0bc2e6bfd85ae22bd/src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs#L354-L358 and https://github.com/dotnet/runtime/blob/8761cb9cf2201054409dd8e0bc2e6bfd85ae22bd/src/mono/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.Mono.cs#L157-L165).

The intent is to formalize the existing metric APIs for the JIT, add a JIT time metric, and make them public. Users can query this information in-process, in real-time without collecting events via `EventPipe` out-of-processs or `EventListener` in-process.

## Proposed API

```csharp
namespace System.Runtime.CompilerServices
{
    public struct JitInfo
    {
        public long CompiledILBytes { get; }
        public long CompiledMethodCount { get; }
        public TimeSpan CompilationTime { get; }
    }

    public partial class RuntimeHelpers
    {
        public static JitInfo GetJitInfo();
        public static JitInfo GetJitInfoForCurrentThread();
    }
}
```

## Usage Examples

These can be used (as the internal ones already are) to produce an `EventCounter`:
```csharp
_ilBytesJittedCounter ??= new PollingCounter("il-bytes-jitted", this, () => System.Runtime.CompilerServices.RuntimeHelpers.GetJitInfo().CompiledBytes) { DisplayName = "IL Bytes Jitted", DisplayUnits = "B" };
```

A user whose application runs third-party code with potentially unknown performance characteristics, could periodically check this information to better understand the behavior of this code. They could use this as follows:
```csharp
using System.Runtime.CompilerServices;

// ...

TimeSpan before = RuntimeHelpers.GetJitInfoForThread().CompilationTime;

ThirdPartyCode.DoWork();

TimeSpan after = RuntimeHelpers.GetJitInfoForThread().CompilationTime;

if ((after - before) > threshold)
	ThirdPartyCode.Unload();

// ...
```

## Alternative Designs

```csharp
namespace System.Runtime.CompilerServices.RuntimeHelpers
{
    public static class JitInfo
    {
        public static long GetCompiledILBytes();
		public static long GetCompiledILBytesForCurrentThread();
        public static long GetCompiledMethodCount();
		public static long GetCompiledMethodCountForCurrentThread();
        public static TimeSpan GetCompilationTime();
		public static TimeSpan GetCompilationTimeForCurrentThread();
    }
}
```

There is also a combo of events named `MethodJittingStarted_V1` and `MethodLoadVerbose_V2` which signify the start and stop of the JIT compiling a method. These events can be turned on and listened to in-process using `EventListener`. You could theoretically aggregate these same statistics using these events. There would be some latency inherent to that design, however, since native events are buffered inside EventPipe and getting up-to-date information is dependent on the `EventListener.OnEventWritten` callback being scheduled on the CPU.

## Risks

There are few risks worth noting:

* Collecting the data for `JitInfo.CompilationTime` will most likely require using the QPC APIs in the runtime PAL. Windows documentation specifies these APIs can come with <300 ns latency. Other platforms and architectures may have different performance expectations for the equivalent APIs. This latency is going to be dominated by the time spent actually JITting, though.
* Is the `ForThread` variant useful?
  * There is no guarantee that the JIT needs to compile a method on the same thread that is going to execute it. e.g., Tiering up happens on a background thread, reJIT via a profiler, On Stack Replacement (OSR) potentially in the future, or MultiCoreJit (though that is mostly turned off if I recall).
  * Should we consider JITting on a specific thread to be an implementation detail?
* Returning a `struct` in the managed API means that a user could observe tearing when reading the data. The JIT updates individual fields atomically, meaning a user could theoretically observe `CompiledILBytes` to increment before `CompiledMethodCount` if `GetJitInfo` is called at precisely the right moment in time.
  * This shouldn't pose an issue, though. Users that are using this API are most likely power users who will read the documentation and see a note about potential tearing. To observe the tearing, a user would need to be polling this struct as fast or faster than the JIT can JIT a method (since the native values are updated at the end of a JIT). This would mean a user is calling `GetJitInfo` in a tight loop with less than a millisecond between calls in some cases. This would consume quite a bit of CPU and most likely not be a use case. Otherwise, a user would simply see a point in time value with potentially some skew from the tearing.
  * Tearing would only be relevant for the global values from `GetJitInfo` and not `GetJitInfoForThread`.

CC @tommcdon @tannergooding @BruceForstall @benmwatson

**Comment by teo-tsirpanis** [comment]
Tearing could be resolved by wrapping a single internal class inside the `JitInfo` struct, just like how `GCMemoryInfo` is implemented. But depending on the potentially different performance requirements of this API this could be unacceptable. Or was `GCMemoryInfo` implemented that way only to prevent memory safety issues of the `PauseDurations` and `GenerationInfo` spans it returns?

**Comment by reflectronic** [comment]
`GCMemoryInfo` is a struct because it originally only had a handful of fields. Then they decided to add many more fields, and they did not want to make a massive struct behemoth, so they needed to essentially make it a class without changing `GCMemoryInfo` itself. I think the API review team will want this to be a class to avoid the same mistake.

**Comment by josalem** [comment]
> Tearing could be resolved by wrapping a single internal class inside the JitInfo struct

To clarify, are you suggesting we create a class that has these values on it and retrieve all the values at once, atomically using a pointer to this class?

Or are you suggesting we replace the `struct JitInfo` with something like:

```csharp
static public class JitInfo
{
    static public long GetCompiledILBytes();
    ...
}
```

Also, CC @lateralusX & @steveisok since there will be a small amount of work in the Mono runtime required to light up the `CompilationTime` value.

**Comment by teo-tsirpanis** [comment]
As @reflectronic said, `GCMemoryInfo` was made a struct because it initially had few fields and started wrapping an internal class when it became bigger (and when one of its properties was a span), and `JitInfo` should better become a class from the start.

**Comment by lateralusX** [comment]
In Mono we have internal JIT stats that can be retrieve (currently dumped into console or file on shutdown) and I know there have been requests to get this info into some sort of API as well (or part of some EventSource) for better monitoring capabilities, it includes a lot of data and some runtime internals, but I'm just include it here for reference as JIT data that could be of interest:

```
                g_print ("Mono Jit statistics\n");
		g_print ("Max code size ratio:    %.2f (%s)\n", mono_jit_stats.max_code_size_ratio / 100.0,
				 mono_jit_stats.max_ratio_method);
		g_print ("Biggest method:         %" G_GINT32_FORMAT " (%s)\n", mono_jit_stats.biggest_method_size,
				 mono_jit_stats.biggest_method);

		g_print ("Delegates created:      %" G_GINT32_FORMAT "\n", mono_stats.delegate_creations);
		g_print ("Initialized classes:    %" G_GINT32_FORMAT "\n", mono_stats.initialized_class_count);
		g_print ("Used classes:           %" G_GINT32_FORMAT "\n", mono_stats.used_class_count);
		g_print ("Generic vtables:        %" G_GINT32_FORMAT "\n", mono_stats.generic_vtable_count);
		g_print ("Methods:                %" G_GINT32_FORMAT "\n", mono_stats.method_count);
		g_print ("Static data size:       %" G_GINT32_FORMAT "\n", mono_stats.class_static_data_size);
		g_print ("VTable data size:       %" G_GINT32_FORMAT "\n", mono_stats.class_vtable_size);
		g_print ("Mscorlib mempool size:  %d\n", mono_mempool_get_allocated (mono_defaults.corlib->mempool));

		g_print ("\nInitialized classes:    %" G_GINT32_FORMAT "\n", mono_stats.generic_class_count);
		g_print ("Inflated types:         %" G_GINT32_FORMAT "\n", mono_stats.inflated_type_count);
		g_print ("Generics virtual invokes: %ld\n", mono_jit_stats.generic_virtual_invocations);

		g_print ("Sharable generic methods: %" G_GINT32_FORMAT "\n", mono_stats.generics_sharable_methods);
		g_print ("Unsharable generic methods: %" G_GINT32_FORMAT "\n", mono_stats.generics_unsharable_methods);
		g_print ("Shared generic methods: %" G_GINT32_FORMAT "\n", mono_stats.generics_shared_methods);
		g_print ("Shared vtype generic methods: %" G_GINT32_FORMAT "\n", mono_stats.gsharedvt_methods);

		g_print ("IMT tables size:        %" G_GINT32_FORMAT "\n", mono_stats.imt_tables_size);
		g_print ("IMT number of tables:   %" G_GINT32_FORMAT "\n", mono_stats.imt_number_of_tables);
		g_print ("IMT number of methods:  %" G_GINT32_FORMAT "\n", mono_stats.imt_number_of_methods);
		g_print ("IMT used slots:         %" G_GINT32_FORMAT "\n", mono_stats.imt_used_slots);
		g_print ("IMT colliding slots:    %" G_GINT32_FORMAT "\n", mono_stats.imt_slots_with_collisions);
		g_print ("IMT max collisions:     %" G_GINT32_FORMAT "\n", mono_stats.imt_max_collisions_in_slot);
		g_print ("IMT methods at max col: %" G_GINT32_FORMAT "\n", mono_stats.imt_method_count_when_max_collisions);
		g_print ("IMT trampolines size:   %" G_GINT32_FORMAT "\n", mono_stats.imt_trampolines_size);

		g_print ("JIT info table inserts: %" G_GINT32_FORMAT "\n", mono_stats.jit_info_table_insert_count);
		g_print ("JIT info table removes: %" G_GINT32_FORMAT "\n", mono_stats.jit_info_table_remove_count);
		g_print ("JIT info table lookups: %" G_GINT32_FORMAT "\n", mono_stats.jit_info_table_lookup_count);

```

There is also a number of EventPipe events that currently cover JIT stats, some already mentioned above, the method load and JIT start events, but there is also events around inlining, failed inlining, tiered compilation, allocated code buffers etc that could interesting to monitor.

**Comment by noahfalk** [comment]
> System.Runtime.CompilerServices.RuntimeHelpers

This may be a surprising place for this new type to show up. RuntimeHelpers is described in the docs as "Provides a set of static methods and properties that provide support for compilers." I don't know if the BCL review folks will have a good alternate suggestion? For example GC is directly in System. 

> Should we consider JITting on a specific thread to be an implementation detail?

I would define it as whichever managed thread is directly waiting for the JIT to complete, regardless whether the JIT implementation executes inline on that thread.

> A user whose application runs untrusted third-party code

I would describe this as a user who wants to monitor or profile some code that they **DO** have some trust in. In the example above the developer would need to trust that code doesn't hang, crash, steal data, or try to hide all the jitting by transferring control to a different thread. If the third party code was written by someone with malicious intent they could easily circumvent the measurement.

> GetJitInfoForThread()

For consistency GC uses "ForCurrentThread" ([GC.GetAllocatedBytesForCurrentThread](https://docs.microsoft.com/en-us/dotnet/api/system.gc.getallocatedbytesforcurrentthread?view=net-5.0))

> public int CompiledILBytes { get; set; }

Are the public setters necessary?

**Comment by tannergooding** [comment]
I think that `RuntimeHelpers` is probably the best place for it at this time. Other specialized methods related to the JIT live here already.

**Comment by terrajobst** [comment]
* It seems we can't meaningfully return information on a per-thread basis
    - If we do, we'd add a parameter rather than a new set of methods, probably defaulted such that it would return the global state when called without specifying.
* We should probably put this in `System.Runtime` as `System.Runtime.CompilerServices` is largely for APIs we expect to be used by compilers (or by people who want write very low level code). These seem on par with GC information APIs.
* Let's expose them as static methods so if we need to expose more information that needs snapshot semantics, we can make it struct/class and add corresponding instance properties.

```C#
namespace System.Runtime
{
    public static class JitInfo
    {
        public static long GetCompiledILBytes();
        public static long GetCompiledMethodCount();
        public static TimeSpan GetCompilationTime();
    }
}
```

**Comment by krwq** [comment]
@josalem can this issue be closed now since the PR is merged?

**Comment by josalem** [comment]
Yes, I must have missed putting a closing verb on the PR.

Closed via #55046
